waiter/src/waiter/auth/composite.clj

```diff
@@ -0,0 +1,69 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.composite
+  (:require [clojure.string :as string]
+            [plumbing.core :as pc]
+            [waiter.auth.authentication :as auth]
+            [waiter.util.utils :as utils]))
+
+(defrecord CompositeAuthenticator [provider-name->authenticator default-authentication-provider]
+  auth/Authenticator
+
+  (get-authentication-providers [_]
+    (keys provider-name->authenticator))
+  (process-callback [_ {{:keys [authentication-provider]} :route-params :as request}]
+    (if-let [authenticator (get provider-name->authenticator authentication-provider)]
+      (auth/process-callback authenticator request)
+      (throw (ex-info (str "Unknown authentication provider " authentication-provider)
+                      {:available-authenticators (keys provider-name->authenticator)
+                       :authentication-provider authentication-provider
+                       :status 400}))))
+  (wrap-auth-handler [_ request-handler]
+    (let [provider-name->handler (pc/map-vals #(auth/wrap-auth-handler % request-handler) provider-name->authenticator)]
+      (fn composite-authenticator-handler [request]
+        (let [authentication (or (get-in request [:waiter-discovery :service-parameter-template "authentication"])
+                                 default-authentication-provider)]
```

Comment:
The default should come from the `service-description-defaults`

Reply:
If I'm not mistaken, `service-parameter-template` contains values from the token. It may be missing `authentication`, in which case the final service description for the request will end up having the value coming from the `service-description-defaults`. Thinking about it, that then means that the default for the composite scheduler should be configured from the `authentication` in `service-description-defaults`.

Reply:
ok, so add "service-description-default-authentication" to the Authenticator context?

Reply:
or `service-description-defaults`. The factory function can extract what it needs.

Reply:
are you fine with just service-description-default-authentication? no need to expose extra

Reply:
okay, it can just be `default-authentication`